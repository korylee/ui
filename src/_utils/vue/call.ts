export function call<A extends any[]>(
  funcs: Function[] | Function,
  ...args: A
): void {
  if (Array.isArray(funcs)) funcs.forEach((func) => call(func, ...args))
  else return funcs(...args)
}

export type MaybeArray<T> = T | T[]
