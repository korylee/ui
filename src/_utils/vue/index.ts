export type { MaybeArray } from './call'

export { default as createInjectionKey } from './create-injection-key'
export { call } from './call'
export { keyof } from './keyof'
export { resolveWrappedSlot } from './resolve-slot'
