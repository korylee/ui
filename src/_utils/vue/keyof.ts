export const keyof = <T>(obj: T): Array<keyof T> => Object.keys(obj) as any
