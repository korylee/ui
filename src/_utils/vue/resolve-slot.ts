import {
  isVNode,
  Slot,
  VNodeArrayChildren,
  Comment,
  Fragment,
  VNodeChild
} from 'vue'

function ensureValidVNode(vnodes: VNodeArrayChildren) {
  const valid = vnodes.some((child) => {
    if (!isVNode(child)) return true
    if (child.type === Comment) return false
    if (
      child.type === Fragment &&
      !ensureValidVNode(child.children as VNodeArrayChildren)
    ) {
      return false
    }
    return true
  })
  if (valid) return vnodes
  return null
}

export const resolveSlot = <T>(slot: Slot | undefined, props?: T) =>
  (slot && ensureValidVNode(slot(props))) ?? null

export const isSlotEmpty = (slot: Slot | undefined) =>
  !(slot && ensureValidVNode(slot()))

export function resolveWrappedSlot<T, R = any>(
  slot: Slot | undefined,
  wrapper: (children: VNodeArrayChildren | null) => T,
  props?: R
) {
  const children = resolveSlot(slot, props)
  return wrapper(children)
}
