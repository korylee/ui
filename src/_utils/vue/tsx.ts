import type { ExtractPropTypes, HTMLAttributes } from 'vue'

export type TsxComponent<
  T extends Record<string, unknown>,
  R extends HTMLAttributes
> = new () => {
  $props: Partial<ExtractPropTypes<T> & Omit<R, keyof T>>
}
