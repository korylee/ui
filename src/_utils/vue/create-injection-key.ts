import { InjectionKey } from 'vue'

export default function createInjectionKey<T> (key: string): InjectionKey<T> {
  return key as any
}
