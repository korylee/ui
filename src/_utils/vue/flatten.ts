import { createTextVNode, Fragment, VNode, VNodeChild } from 'vue'

export function flatten(
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    else if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    } else if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    } else if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children))
        flatten(vNode.children, filterCommentNode, result)
      else flatten([vNode.children as VNodeChild], filterCommentNode, result)
    } else if (vNode.type === Comment && filterCommentNode) return
    else result.push(vNode)
  })
  return result
}
