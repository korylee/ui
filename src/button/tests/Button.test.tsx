import { describe, it, vi, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { Button } from '../index'
import { h } from 'vue'

describe('k-button', () => {
  it('should work with import on demand', () => {
    const inst = mount(Button)
    inst.unmount()
  })
  it('clickable', async () => {
    const onClick = vi.fn()
    const inst = mount(Button, { props: { onClick } })
    expect(onClick).not.toHaveBeenCalled()
    await inst.trigger('click')
    expect(onClick).toHaveBeenCalled()
    inst.unmount()
  })
  it('keyboard', async () => {
    const onClick = vi.fn()
    const inst = mount(Button, { props: { keyboard: true, onClick } })
    await inst.trigger('click')
    expect(onClick).toBeCalledTimes(1)
    await inst.trigger('keydown.space')
    expect(onClick).toBeCalledTimes(1)
    await inst.trigger('keydown.enter')
    expect(onClick).toBeCalledTimes(1)
    inst.unmount()
  })
  it('disabled', async () => {
    const onClick = vi.fn()
    const inst = mount(Button, { props: { disabled: true, onClick } })
    await inst.trigger('click')
    expect(onClick).not.toHaveBeenCalled()
    inst.unmount()
  })
  it('pressed native event & attr tsx type check', () => {
    ;<div>
      <Button onMousedown={() => {}}></Button>
      <Button formaaction=""></Button>
    </div>
  })
  it('show work with `attr-type prop`', () => {
    ;['button', 'submit', 'reset'].forEach((type) => {
      const inst = mount(Button, { props: { attrType: type } })
      expect(inst.find('button').attributes('type')).toContain(type)
      inst.unmount()
    })
  })
  it(`show work with 'size' prop`, () => {
    ;['tiny', 'small', 'medium', 'large'].forEach((size) => {
      const inst = mount(Button, { props: { size } })
      expect(inst.find('button').attributes('style')).toMatchSnapshot()
      inst.unmount()
    })
  })
  it('should work iwth `test` prop', () => {
    const inst = mount(Button, {
      props: { text: true },
      slots: { default: () => 'test' }
    })

    expect(inst.find('button').attributes('style')).toMatchSnapshot()
    expect(inst.find('.k-button__content').element.textContent).toBe('test')
    inst.unmount()
  })
})
