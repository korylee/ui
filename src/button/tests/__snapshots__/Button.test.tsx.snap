// Vitest Snapshot v1

exports[`k-button > should work iwth \`test\` prop 1`] = `"--k-bezier: cubic-bezier(.4, 0, .2, 1); --k-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --k-ripple-duration: .6s; --k-wave-opacity: 0.6; --k-opacity-disabled: 0.5; --k-font-weight: 400; --k-width: initial; --k-height: initial; --k-font-size: 14px; --k-padding: initial; --k-border-radius: initial; --k-icon-size: 18px; --k-icon-margin: 6px; --k-color: #0000; --k-color-hover: #0000; --k-border: none; --k-border-hover: none;"`;

exports[`k-button > show work with 'size' prop 1`] = `"--k-bezier: cubic-bezier(.4, 0, .2, 1); --k-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --k-ripple-duration: .6s; --k-wave-opacity: 0.6; --k-opacity-disabled: 0.5; --k-font-weight: 400; --k-width: initial; --k-height: 22px; --k-font-size: 12px; --k-padding: 0 6px; --k-border-radius: 3px; --k-icon-size: 14px; --k-icon-margin: 6px; --k-color: #0000; --k-color-hover: #0000; --k-color-pressed: #0000; --k-border: 1px solid rgb(224, 224, 300);"`;

exports[`k-button > show work with 'size' prop 2`] = `"--k-bezier: cubic-bezier(.4, 0, .2, 1); --k-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --k-ripple-duration: .6s; --k-wave-opacity: 0.6; --k-opacity-disabled: 0.5; --k-font-weight: 400; --k-width: initial; --k-height: 28px; --k-font-size: 14px; --k-padding: 0 10px; --k-border-radius: 3px; --k-icon-size: 18px; --k-icon-margin: 6px; --k-color: #0000; --k-color-hover: #0000; --k-color-pressed: #0000; --k-border: 1px solid rgb(224, 224, 300);"`;

exports[`k-button > show work with 'size' prop 3`] = `"--k-bezier: cubic-bezier(.4, 0, .2, 1); --k-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --k-ripple-duration: .6s; --k-wave-opacity: 0.6; --k-opacity-disabled: 0.5; --k-font-weight: 400; --k-width: initial; --k-height: 34px; --k-font-size: 14px; --k-padding: 0 14px; --k-border-radius: 3px; --k-icon-size: 18px; --k-icon-margin: 6px; --k-color: #0000; --k-color-hover: #0000; --k-color-pressed: #0000; --k-border: 1px solid rgb(224, 224, 300);"`;

exports[`k-button > show work with 'size' prop 4`] = `"--k-bezier: cubic-bezier(.4, 0, .2, 1); --k-bezier-ease-out: cubic-bezier(0, 0, .2, 1); --k-ripple-duration: .6s; --k-wave-opacity: 0.6; --k-opacity-disabled: 0.5; --k-font-weight: 400; --k-width: initial; --k-height: 40px; --k-font-size: 15px; --k-padding: 0 18px; --k-border-radius: 3px; --k-icon-size: 20px; --k-icon-margin: 6px; --k-color: #0000; --k-color-hover: #0000; --k-color-pressed: #0000; --k-border: 1px solid rgb(224, 224, 300);"`;
